Feature/OT111-34 Create Endpoint Update Usuario.

HugoLucero opened the pull request:
**LINK JIRA**
https://alkemy-labs.atlassian.net/jira/software/c/projects/OT111/boards/168?modal=detail&selectedIssue=OT111-34

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------

- [x] Se agregaron los métodos del servicio correspondientes (findById y updateUser)
- [x] Se creó un nuevo DTO para el update (pensado en caso de que se quiera modificar la contraseña)
- [x] Se crearon los endpoints de get user y put user
- [x] Se realizaron pruebas y funcionó ok.
- [x] Some bugs fixes.
- [x] Se arreglaron conflictos y se resolvieron temas de importaciones. 
- [x] Se modifico el path sobre el security config ya que debía agregar el id a users ("/users/**")
- [x] Se probaron los cambios, hasta el de codificar el password en caso de que se requiera
- [x] Se modificó el retorno del PUT porque para prueba se usó que devolviera el dto y asi poder verificar su funcionamiento.

denisdimarco committed:
OT111-35 - Created UserController, implemented GET /users

denisdimarco committed:
OT111-35 - Created UserService, DTOs, Mappers.

Created UserService & UserServiceImpl, implemented findAllUsers method.
Created RoleDTO & UserDTO
Created AbstractMapper, UserMapper & RoleMapper.

denisdimarco committed:
OT111-35 - Fix some inconsistencies

build.gradle: removed duplicated dependency.

NewsEntity : removed completed TODOs comments

UserEntity: Added missing @Entity annotation  and @NoArgsConstructor, @Builder, also modified 'Role' name to 'RoleEntity'.

ActivityRepository: Added missing @Repository annotation.

RoleRepository: Added missing @Repository annotation.

UserRepository: Added missing @Repository annotation.

HugoLucero committed:
Merge remote-tracking branch 'origin/feature/OT111-35' into feature/OT111-34

HugoLucero committed:
OT111 - 34 - Cracion de endpoint PATCH y metodos update, etc

Crissil07 committed:
Novedades - Merge con develop

HugoLucero committed:
OT111-34 - Replace PATH/PUT mapping, some bugs were fixed.

Crissil07 committed:
News controller/mapper/DTO/Service/ServiceImpl - prueba con security

HugoLucero committed:
Merge branch 'develop' into feature/OT111-34

OsvaldoGWP commented on build.gradle:
```diff
@@ -19,7 +19,6 @@ repositories {
 }
 
 dependencies {
-	implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
```
antes de q pueda revisar el pr porfavor resolver los conflictos al actualizar con develop

Crissil07 committed:
Creacion CategoryDTO/Mapper - prueba con postman funcionando

Crissil07 committed:
Agregado antMachers - news - hasRole("Admin")

HugoLucero committed:
OT111-34 - Se Resolvieron algunos conflictos y optimizaron importaciones.

HugoLucero committed:
OT111 - 34 - Se modifico el retorno del controller

HugoLucero committed:
OT111 - 34 - Resolve conflicts.

HugoLucero committed:
OT111-34 - Resolve conflicts.

HugoLucero committed:
OT111-34 - Resolve conflicts.

OsvaldoGWP commented on src/main/java/com/alkemy/ong/controller/UserController.java:
```diff
@@ -20,10 +17,26 @@
 
     @GetMapping
-    public ResponseEntity<List<UserDTO>> getAllUsers() {
+    public ResponseEntity<?> getAllUsers() {
```
lo q se borro estaba correcto

OsvaldoGWP commented on src/main/java/com/alkemy/ong/controller/UserController.java:
```diff
@@ -20,10 +17,26 @@
 
-//TODO Pull Request para agregar Path /users en SecurityConfiguration para que lo pueda solo trabajar un admin
+    @GetMapping("/{id}")
+    public ResponseEntity<?> getUser(@PathVariable Long id){
```
no usar signo de pregunta "?" sino colocar el tipo de respuesta en este caso UserDto

OsvaldoGWP commented on src/main/java/com/alkemy/ong/service/UserService.java:
```diff
@@ -1,9 +1,14 @@
+
+    UserDTO findById(Long id);
+
+    UserUpdateDTO updateUser(UserUpdateDTO UserUpdateDTO, Long id);
```
este metodo q devuelva un void ya q no se no es necesario una respuesta en este caso

OsvaldoGWP replied:
normalmente los put o patch devuelven un codigo 204 que dice q no tiene cuerpo

OsvaldoGWP commented on src/main/java/com/alkemy/ong/service/impl/UserServiceImpl.java:
```diff
@@ -1,28 +1,69 @@
+
+            userEntity.get().setFirstName(userUpdateDTO.getFirstName());
+            userEntity.get().setLastName(userUpdateDTO.getLastName());
+        if (!userUpdateDTO.getPassword().isEmpty()){
```
identar codigo este if esta a la misma altura q su if contenedor

OsvaldoGWP commented on src/main/java/com/alkemy/ong/service/impl/UserServiceImpl.java:
```diff
@@ -1,28 +1,69 @@
+    @Transactional
+    public UserUpdateDTO updateUser(UserUpdateDTO userUpdateDTO, Long id) {
+        Optional<UserEntity> userEntity = userRepository.findById(id);
+        if(!userEntity.isEmpty()){
```
una mejora sutil seria invertir este if osea  if(userEntity.isEmpty()) return null;
esto hace mejor legible el codigo

HugoLucero replied:
repare anteriormente que el método update fuera void ya que no devolvía nada y al hacerlo no podría devolver un null, lo único que se me ocurre sería: "(userEntity != null){...}". Estuve revisando algunas mejores prácticas sobre los Optional y encontré casi lo mismo (userEntity != null).

https://refactorizando.com/java-optional-buenas-practicas/

![Captura de pantalla de 2021-12-09 20-36-29](https://user-images.githubusercontent.com/90858496/145492389-a38ebba7-4ec0-45f7-8106-a1fc98eca62a.png)

OsvaldoGWP replied:
claro al ser un void no deberia retornar nada, no deberia estar la clausula return

Crissil07 committed:
Merge branch 'develop' into feature/OT111-45

HugoLucero committed:
OT111 - 34 - Se actualizó según revisión.

Crissil07 committed:
Cambios del PR REALIZADOS y TESTEADO

Crissil07 committed:
Corregido

OsvaldoGWP commented on src/main/java/com/alkemy/ong/controller/UserController.java:
```diff
@@ -23,7 +22,23 @@
 
+    @PutMapping("/{id}")
+    public ResponseEntity<UserUpdateDTO> updateUser(@RequestBody UserUpdateDTO userUpdateDTO, @PathVariable Long id){
+        UserDTO userDTO = userService.findById(id);
```
guiarse de category el update

OsvaldoGWP commented on src/main/java/com/alkemy/ong/controller/UserController.java:
```diff
@@ -23,7 +22,23 @@
-//TODO Pull Request para agregar Path /users en SecurityConfiguration para que lo pueda solo trabajar un admin
+    @GetMapping("/{id}")
+    public ResponseEntity<UserDTO> getUser(@PathVariable Long id){
+        UserDTO userDTO = userService.findById(id);
```
guiarse del crud de category q esta en develop ya esta implementado con manejo de errores

OsvaldoGWP commented on src/main/java/com/alkemy/ong/repository/RoleRepository.java:
```diff
@@ -4,7 +4,7 @@
 import org.springframework.stereotype.Repository;
 
-@Repository
+@Repository //IMPORTANT - Added @Repository annotation while working on OT111-35
```
No hace falta esta anotacion, chequear este link 
https://www.it-swarm-es.com/es/java/repository-no-es-necesario-al-implementar-jparepository/832385271/#:~:text=De%20hecho%2C%20no%20es%20necesario,de%20las%20interfaces%20predefinidas%20Repository%20.&text=Esto%20buscar%C3%A1%20Spring%20Data%20repositories,no%20est%C3%A1n%20anotados%20con%20%40Repository%20.

HugoLucero replied:
Genial Osvaldo... No era mío, pero ahí lo borre.

OsvaldoGWP commented on src/main/java/com/alkemy/ong/service/impl/UserServiceImpl.java:
```diff
@@ -1,28 +1,67 @@
+import java.util.Optional;
 
 @Service
 public class UserServiceImpl implements UserService {
```
Chequiemos el servicio de category, ya q ese seria nuestra guia para de un crud, ya esta implementado el manejo de excepciones globales
 cualquier cosa consultar

OsvaldoGWP commented on src/main/java/com/alkemy/ong/controller/UserController.java:
```diff
@@ -23,7 +22,23 @@
     public ResponseEntity<List<UserDTO>> getAllUsers() {
```
Actualizar con develop y resolver conflictos

HugoLucero committed:
Update from develop

HugoLucero committed:
OT111-34 - Merge & resolve conflicts. The changes seen in the pr.

HugoLucero commented:
Dejo captura de funcionamiento ok.
![Captura de pantalla de 2021-12-12 20-37-57](https://user-images.githubusercontent.com/90858496/145734476-a5f44199-81ec-47e9-a26a-c42ee4adb887.png)

Crissil07 committed:
En correccion

Crissil07 committed:
Merge branch 'develop' into feature/OT111-45

# Conflicts:
#	src/main/java/com/alkemy/ong/auth/config/SecurityConfiguration.java
#	src/main/java/com/alkemy/ong/model/mapper/CategoryMapper.java

OsvaldoGWP commented on src/main/java/com/alkemy/ong/model/mapper/UserMapper.java:
```diff
@@ -35,4 +31,22 @@ public UserDTO entity2DTO(UserEntity entity) {
+                .id(entity.getId())
+                .firstName(entity.getFirstName())
+                .lastName(entity.getLastName())
+                .password(entity.getPassword())
```
No devolver el password es un dato sensible

HugoLucero replied:
Aaa ok. Yo lo puse xq pensé que al actualizar el usuario podría querer actualizar el password y como es sobre la conversión del dto a entity, eso devuelve la entidad, pero el método del servicio es un void, en teoría al front no enviaría nada, solo lo haría al repositorio. Y consulta como se envía la actualización del password en caso de se quisiera modificar?

OsvaldoGWP replied:
sisi por eso en el comentario de abajo puse q este metodo no se esta usando osea sacarlo si no se usa, normalmente se tiene un endpoint especifico para actualizar la contraseña

OsvaldoGWP commented on src/main/java/com/alkemy/ong/model/mapper/UserMapper.java:
```diff
@@ -35,4 +31,22 @@ public UserDTO entity2DTO(UserEntity entity) {
     }
-}
+
+    public UserUpdateDTO entityUpdate2DTO(UserEntity entity) {
```
me parece q no se esta usando este metodo

Crissil07 replied:
Se dejo comentado para ser verificado, independientemente se resolvio el conflicto.

OsvaldoGWP commented on src/main/java/com/alkemy/ong/model/request/user/UserUpdateDTO.java:
```diff
@@ -0,0 +1,32 @@
+public class UserUpdateDTO {
+
+
+    private Long id; //Check if needed.
```
esto no iria, lo mismo private String modifiedDate;

    private String modifiedBy; tampoco iria

OsvaldoGWP commented on src/main/java/com/alkemy/ong/model/request/user/UserUpdateDTO.java:
```diff
@@ -0,0 +1,32 @@
+
+    private Long id; //Check if needed.
+
+    private String firstName;
```
validar los campos q son requeridos para actualizar un usuario, con anotaciuones por ejemplo @NotEmpty

Crissil07 replied:
Resuelto

OsvaldoGWP commented on src/main/java/com/alkemy/ong/model/request/user/UserUpdateDTO.java:
```diff
@@ -0,0 +1,32 @@
+
+    private String photo;
+
+    private List<RoleDTO> roles;
```
aqui quizas deberiamos recibir una lista de enteros que serian los ids de los roles

OsvaldoGWP replied:
@Crissil07

Crissil07 replied:
Cambiada por una lista de Integers

Crissil07 committed:
Corregido el PR 13/12

Crissil07 committed:
Corregido Metodo Get.

Crissil07 committed:
Corregido y funcionando PR

Crissil07 committed:
Completando tarea 34(cris)

# Conflicts:
#	src/main/java/com/alkemy/ong/auth/config/SecurityConfiguration.java

Crissil07 committed:
Tarea 34 - Testeada y funcionando. Realizadas las correcciones y finalizada

Crissil07 committed:
Tareas corregidas - 34

Crissil07 commented:
Conflictos resueltos. Revisar.

OsvaldoGWP committed:
Se saca pssword del request para actualizar usuario
